pkg/api/helpers_test.go

```diff
@@ -397,6 +398,275 @@ func TestMatchTaint(t *testing.T) {
 	}
 }
 
+func TestTolerationToleratesTaint(t *testing.T) {
+	genForgivenessSeconds := func(f int64) *int64 {
+		return &f
+	}
+
+	testCases := []struct {
+		description     string
+		toleration      Toleration
+		taint           Taint
+		expectTolerated bool
+	}{
+		{
+			description: "toleration and taint have the same key and effect, and operator is Exists, and taint has no value, expect tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpExists,
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: true,
+		},
+		{
+			description: "toleration and taint have the same key and effect, and operator is Exists, and taint has some value, expect tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpExists,
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "bar",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: true,
+		},
+		{
+			description: "toleration and taint have the same key, effect and value, and operator is Equal, expect tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpEqual,
+				Value:    "bar",
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "bar",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: true,
+		},
+		{
+			description: "toleration and taint have the same key and effect, but different values, and operator is Equal, expect not tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpEqual,
+				Value:    "value1",
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "value2",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: false,
+		},
+		{
+			description: "toleration and taint have the same key and value, but different effects, and operator is Equal, expect not tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpEqual,
+				Value:    "bar",
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "bar",
+				Effect: TaintEffectNoExecute,
+			},
+			expectTolerated: false,
+		},
```

Comment:
@ingvagabund Previously we had something like what you are suggesting here -- the choices were NoSchedule, NoScheduleNoAdmit, NoScheduleNoAdmitNoExecute, for the reason you mentioned. But we decided to split out NoExecute when introducing forgiveness, because as part of forgiveness we make it possible to represent various node problems using taints, and allow a pod to use a toleration with forgiveness to say it is willing to stay bound to a node for a certain amount of time once the taint appears. But such a pod may not want to schedule onto the node when the taint is already there. If the only way a pod can say "keep me bound to the node for X seconds when taint T appears" is for it to add a NoScheduleNoAdmitNoExecute toleration for T (with forgivenessSeconds X), then it is also forced to tolerate getting scheduled onto the node, which it may not want.

So we decided to split out NoExecute. This means if you want the "keep me bound to the node for X seconds when taint T appears" behavior, then you tolerate NoExecute for T with forgivenessSeconds X, and independently you can decide whether or not you want to schedule onto a node where T is already present (by deciding whether to have a NoSchedule toleration for T).

Does that make sense?

It does mean that to represent a machine problem we need to add both a NoExecute taint and a NoSchedule taint at the same time.

(Note to @kevin-wangzefeng : we may need to kill when a taint is added, when a taint is updated, when a pod is scheduled, when a pod is updated). 

[For the sake of completeness, I will say that what I wrote above is a slight lie, in that we could keep the old set of taint effects (that has the `<` property you mentioned) and still be able to distinguish the cases I mentioned. But it would be confusing for the user. It would work like this: If a node has both a NoSchedule taint and a NoScheduleNoAdmitNoExecute taint and you tolerate both taints, then you can schedule there and stay bound. If a node has both taints and you only tolerate the NoScheduleNoAdmitNoExecute taint, then you wouldn't be able to schedule there because you don't tolerate the NoSchedule taint. Our matching rules between taints and tolerations are such that tolerating a NoScheduleNoAdmitNoExecute taint has no implication for other taints that might also exist on the node.]